### ISSUE TITLE
Continuous integration and nightly builds

### ISSUE THREAD

**Issue by raulk** [libp2p/go-libp2p#621]
**TL;DR: Let's build on the [libp2p workspace](https://github.com/libp2p/workspace-go-libp2p) to provide continuous integration and nightly builds for binaries (right now the daemon, because libp2p is really a library).**

Problems:

* right now we abuse patch releases to integrate module changes upstream; this creates noise, unpredictability, and a hoard of insignificant releases for tiny tweaks in some cases (e.g. just a change of default value in a variable).
* we don't notice breakage across modules until we've made a release and integrated it upstream, which sucks because it's already too late to fix it, and we're left with a bad release, forever etched in the annals of history.
* we have no means to run integration tests across master builds.

Proposal:
* implement a Travis build in workspace-go-libp2p that pulls in all masters across modules, cross-references them via `replace` gomod directives, and runs all tests. 
  * The workspace script already has commands for much of this, we need to cook up a travis config file for this.
* have the above build produce nightly binaries for the daemon, and any other main programs we have.

**Comment by jacobheun** [comment]
> have the above build produce nightly binaries for the daemon, and any other main programs we have.

If this could also upload to ipfs that would be a nice bonus. libp2p/interop currently uses ipfs distribution to pull in the go daemon binary via an npm package, https://github.com/libp2p/npm-go-libp2p-dep. We could potentially upgrade that to do some smarter version checking and do nightly interop builds of the latest releases.

**Comment by Stebalien** [comment]
IMO, we should still cut releases of individual modules so downstream projects can update these modules to get new features. We just shouldn't cut releases of go-libp2p for every tiny change.

**Comment by marten-seemann** [comment]
This is not a problem any more, since go-libp2p is essentially a mono-repo now.
